packages/@aws-cdk/cdk/lib/lazy.ts

```diff
@@ -0,0 +1,166 @@
+import { IResolveContext, Token } from "./token";
+
+/**
+ * Interface for lazy string producers
+ */
+export interface IStringValue {
+  /**
+   * Produce the string value
+   */
+  produce(context: IResolveContext): string | undefined;
+}
+
+/**
+ * Interface for lazy list producers
+ */
+export interface IListValue {
+  /**
+   * Produce the list value
+   */
+  produce(context: IResolveContext): string[] | undefined;
+}
+
+/**
+ * Interface for lazy number producers
+ */
+export interface INumberValue {
+  /**
+   * Produce the number value
+   */
+  produce(context: IResolveContext): number | undefined;
+}
+
+/**
+ * Interface for lazy untyped value producers
+ */
+export interface IAnyValue {
+  /**
+   * Produce the value
+   */
+  produce(context: IResolveContext): any;
+}
+
+/**
+ * Options for creating a lazy string token
+ */
+export interface LazyStringValueOptions {
+  /**
+   * Use the given name as a display hint
+   *
+   * @default - No hint
+   */
+  readonly displayHint?: string;
+}
+
+/**
+ * Options for creating a lazy list token
+ */
+export interface LazyListValueOptions {
+  /**
+   * Use the given name as a display hint
+   *
+   * @default - No hint
+   */
+  readonly displayHint?: string;
+
+  /**
+   * If the produced list is empty, return 'undefined' instead
+   *
+   * @default false
+   */
+  readonly omitEmpty?: boolean;
+}
+
+/**
+ * Options for creating lazy untyped tokens
+ */
+export interface LazyAnyValueOptions {
+  /**
+   * Use the given name as a display hint
+   *
+   * @default - No hint
+   */
+  readonly displayHint?: string;
+
+  /**
+   * If the produced value is an array and it is empty, return 'undefined' instead
+   *
+   * @default false
+   */
+  readonly omitEmptyArray?: boolean;
+}
+
+/**
+ * Lazily produce a value
+ *
+ * Can be used to return a string, list or numeric value whose actual value
+ * will only be calculated later, during synthesis.
+ */
+export class Lazy {
+  public static stringValue(producer: IStringValue, options: LazyStringValueOptions = {}) {
```

Comment:
I kinda would like tho have the return types of those be spelled out on the signature. But I'm cool if you choose not to.